### PR TITLE
Make it possible to cancel a run in the R test explorer

### LINKED ISSUE #15130: It's impossible to cancel a test run in the R test explorer
Extracting from kitchen sink issue [#2636](https://github.com/posit-dev/positron/issues/2636#issuecomment-2767543490):

> And once you start them running, there's no way to stop them? Clicking on the stop button in the test pane doesn't seem to do anything.

and https://github.com/posit-dev/positron/issues/2636#issuecomment-3307352726

> ... its especially problematic if for some reason the tests are stalling/hanging. I've got a project with some particularly tricky code to test, and have had tests just stall forever with only way to stop by exiting positron entirely.

### PR DESCRIPTION
**Pull request by jennybc**
Closes #15130 
@:test-explorer @:web @:win

Adds support for stopping an in-progress test run in the R test explorer. The stop button and `testing.cancelRun` now interrupt the R process we spawn to run `devtools::test()` / `test_active_file()`. The UI also displays accurate statuses on tests/files and reflects that the test run is over.

cancel-test-run-button

There are 2 main pieces to the work:

* Handling the run's cancellation token: Previously we did *nothing* with the token, so R kept running to completion (or not 😬) in the background and the run never truly ended in the UI.

  The R test explorer plugs into the testing framework that Positron inherits from VS Code core. This core *did* mark cancelled tests as "Skipped" after ~10s, but the spinner and timer kept counting up and the test explorer was hung until R finished on its own or the user reloaded the window.

* Interrupting R: This is a lot harder than the cancellation token and requires a direct handle on its process. Previously the test run was spawned through a shell (`spawn(command, { shell: true })`), so the tracked pid was the shell, not R. Now we spawn R directly, so we can knock it on the head. This change caused a few other ripples in the code.

New behaviour on cancel:

* Terminate the R process running the tests:
  - On macOS and linux, we send SIGINT and R unwinds and runs `on.exit` / `withr` / `teardown` cleanup before exiting. Graceful.
  - On Windows, we have to do `taskkill /pid <pid> /T /F`, which just force kills it, without doing cleanup. In theory, it is possible to signal a graceful interrupt, but in reality it's pretty hard and out of scope. `/T` (tree kill) is required because `R.exe` is a front-end that spawns the real interpreter (`Rterm.exe`) as a child, so we must kill the tree.
  - After a 7s grace period, if R hasn't stopped, a non-modal **Force Quit** prompt offers `SIGKILL` (warning that cleanup/teardown may be skipped). Mostly only relevant to the macOS/linux case.
  - In-flight run processes are tracked; if the extension host goes away mid-run (window close/reload), we `SIGKILL` survivors so R can't leak. Very much an edge case.
* The run ends promptly: spinner and timer stop, and you can start another run (instead of the explorer being stuck).
* TEST RESULTS shows "The test run was stopped at the user's request."
* The run's queued-but-not-run items settle to "Skipped".

after-cancel

### Known limitations

- If a package has `Config/testthat/parallel: true` in `DESCRIPTION`, test files still running when you cancel are killed without cleanup. That's just how testthat shuts down its worker subprocesses. It's not something positron-r controls.
- Windows cancel is forceful. `taskkill /F` is `TerminateProcess`, so `on.exit`/teardown do not run. There is no per-process graceful interrupt available to us from Node on Windows and there is no precedent for anything better in VS Code core or any other built-in extension. There are ways to do this with native code, but that would be another adventure like the one re: accessing the Windows registry, which I am not eager to repeat.

More than you ever wanted to know about interrupting R and background tasks: <https://blog.r-project.org/2023/05/23/not-interrupting-background-tasks-with-ctrl-c/>

### Testing

- I worked through a large manual test matrix which I'll put some notes on in a comment. So much of this is interactive and transient UI stuff that is not amenable to non-flaky e2e tests.
- One new e2e test that checks for test actually running, test being interrupted and not completing, test appearing as skipped, and (on macOS/linux) test cleanup happening. I have seen this go from red to green locally and on Windows in CI.
- Update to unit tests of `escapeLabelForRDesc()`.
- At the time of writing, the e2e chromium failures are all related to `New Folder Flow: Python Project › New env: ...` and have no apparent connection to this work.

#### New Features

- The R test explorer now implements "Cancel Test Run".

#### Bug Fixes

- N/A

### Validation Steps

If you want to verify basic functionality, just start a test run in any R package that takes several second or more. Click the square stop button to interrupt it.

Below I'll list the manual test matrix I worked through.

### CODE REVIEW

File: extensions/positron-r/resources/testing/r.pkg.test.explorer.fixture/tests/testthat/test-tricky-desc.R [hunk 1]
```diff
@@ -0,0 +1,24 @@
+test_that("test_that with a
```
Comment (jennybc) [inline extensions/positron-r/resources/testing/r.pkg.test.explorer.fixture/tests/testthat/test-tricky-desc.R:1]: I generalized `test-multi-line-desc.R` to any tricky characters in a test's `desc`. The main work allowed us to simplify the escaping, so I ended up grooming tests about tricky descriptions while I was here.

File: extensions/positron-r/src/testing/runner-testthat.ts [hunk 2]
```diff
@@ -92,25 +130,65 @@ export async function runThatTest(
...
 	LOGGER.info(`Locale info from active R session: ${JSON.stringify(locale, null, 2)}`);
 	let hostFile = '';
-	// TODO @jennybc: if this code stays, figure this out
-	// eslint-disable-next-line no-async-promise-executor
-	return new Promise<string>(async (resolve, reject) => {
-		const childProcess = spawn(command, {
+	return new Promise<string>((resolve, reject) => {
+		const childProcess = spawn(binpath, args, {
 			cwd: wd,
-			shell: true,
```
Comment (jennybc) [inline extensions/positron-r/src/testing/runner-testthat.ts:108]: This is actually a big change: spawning an R process directly vs. inside a spawned shell.

File: extensions/positron-r/src/testing/util-testing.ts [hunk 3]
```diff
@@ -39,15 +39,12 @@ export function uriToFileNodeId(uri: vscode.Uri): string {
...
- * a double-quoted shell `-e` argument, so embedded quotes and backticks must be
- * escaped. The `\n` inside a multi-line description must also be escaped for
- * proper handling on Windows (#10133).
+ * Escape a test label for embedding in a call like
+ * `devtools::test_active_file(..., desc = 'LABEL')`.
  */
 export function escapeLabelForRDesc(label: string): string {
 	return label
-		.replace(/(['"`])/g, '\\$1')
+		.replace(/'/g, '\\\'')
```
Comment (jennybc) [inline extensions/positron-r/src/testing/util-testing.ts:47]: The escaping actually gets simpler, now that we're not using a shell.

**Comment by jennybc** [comment]
Most of the cancellation behaviour is interactive/transient UI that doesn't lend itself to a non-flaky automated test, so I want to record the manual test matrix I've been working with. For review or revisiting or the future.

Platform: I ran locally on macOS.

Good way to observe cleanup: use `withr::defer` / `on.exit` and package
`setup.R` `teardown_env()` to emit a message, which will be visible in the R Language Pack output channel. Or write a sentinel file.

How to check for orphans: use a command like this (change the inner grep as needed):

```
ps -Ao pid,ppid,command | grep -E "bin/exec/R|trap ''" | grep -v grep
```

### Scenarios

- [x] A Baseline graceful cancel. Run a package's tests, cancel mid-run. Expect: R stops promptly; already-run tests keep their results (pass or fail); the in-flight test and any queued-but-un-run tests settle to Skipped; TEST RESULTS shows "The test run was stopped at the user's request."; spinner/timer stop; you can immediately start another run.
- [x] B Cleanup during graceful cancel. A test with in-test `withr::defer()` + `on.exit()`, plus a `setup.R` that registers `teardown_env()` cleanup. Cancel mid-run. Expect: all of it fires (in-test AND package teardown) before R exits. A cancelled run cleans up as thoroughly as a normal one.

    In `tests/testthat/setup.R`:
    ```r
    withr::defer(message(">>> teardown_env defer fired"), teardown_env())
    ```

    In `tests/testthat/test-sleepy.R`:
    ```r
    test_that("cleanup fires on cancel", {
      on.exit(message(">>> on.exit fired"), add = TRUE)
      withr::defer(message(">>> in-test defer fired"))
      message(">>> test running, sleeping now")
      Sys.sleep(30)          # cancel during this window
      expect_true(TRUE)
    })
    ```
    
    In output channel, see:
    ```
    >>> test running, sleeping now
    ! Interrupting test: cleanup fires on cancel
    
    >>> in-test defer fired
    >>> on.exit fired
    >>> teardown_env defer fired
    Execution halted
    ```
- [x] C Force Quit path (unresponsive R). Make R ignore the interrupt by blocking it in a signal-ignoring child inside a test. Cancel. Expect: after ~7s, a non-modal "R is not responding to the interrupt signal... Force Quit" prompt; clicking Force Quit kills R. Caveat: cleanup is skipped (as warned), and a detached `system()` child can survive (reparented to pid 1). Such is life.

    In a test file:
    ```r
    test_that("ignore normal signals to stop", {
      system("sh -c \"trap '' INT TERM; while true; do sleep 1; done\"")
      expect_true(TRUE)
    })
    ```

    Screenshot 2026-07-24 at 5 27 40 PM

    After, in a terminal
    ```
    $ ps -Ao pid,ppid,command | grep -E "bin/exec/R|trap ''" | grep -v grep
    75730     1 sh -c trap '' INT TERM; while true; do sleep 1; done
    $ kill -9 75730
    ```

- [x] D Parallel run (documented limitation). Package has `Config/testthat/parallel: true` in DESCRIPTION; a test writes a START sentinel then a CLEANUP sentinel via `withr::defer`. Cancel mid-run. Expect: files still running show START but no CLEANUP (worker cleanup is lost), and no orphaned worker processes remain. This is testthat's own worker teardown, not something positron-r controls.

    In DESCRIPTION:
    ```
    Config/testthat/parallel: true
    ```

    Example test file (do same for A, B, C, D):
    ```r
    test_that("A slow", {
      cat(sprintf("START file=A pid=%d\n", Sys.getpid()), file = "/tmp/rcancel-cleanup.log", append = TRUE)
      withr::defer(cat(sprintf("CLEANUP file=A pid=%d\n", Sys.getpid()),
                       file = "/tmp/rcancel-cleanup.log", append = TRUE))
      Sys.sleep(30)
      expect_true(TRUE)
    })
    ```

    Look at the log file:
    ```
    $ cat /tmp/rcancel-cleanup.log
    START file=A pid=88238
    START file=B pid=88237
    ```

    Confirm the worker PIDs are not orphaned:
    ```
    $ ps -p 88238,88237 -o pid,ppid,command
      PID  PPID COMMAND
    $ ps -Ao pid,ppid,command | grep -E "bin/exec/R" | grep -v grep
    ```

- [x] E Window close mid-run. Start a run, close the window while it's running. Expect (`ps`): no R process survives (in-flight runs are killed on extension teardown).

    In `tests/testthat/test-sleepy.R`:
    ```r
    test_that("sleepy", {
      cat(sprintf("RUNNER pid=%d\n", Sys.getpid()), file = "SENTINEL", append = TRUE)
      Sys.sleep(60)          # long enough to close the window mid-run
      expect_true(TRUE)
    })
    ```

    While test is running, get PID and behold the runner:
    ```
    .../resources/testing/r.pkg.test.explorer.fixture $ cat tests/testthat/SENTINEL
    RUNNER pid=97094
    .../resources/testing/r.pkg.test.explorer.fixture $ ps -ww -p 97094 -o pid,ppid,command
      PID  PPID COMMAND
    97094 93184 /Library/Frameworks/R.framework/Versions/4.6/Resources/bin/exec/R --no-echo -e devtools::load_all('/Users/jenny/work/positron/extensions/positron-r/resources/testing/vscodereporter');devtools::test('/Users/jenny/work/positron/extensions/positron-r/resources/testing/r.pkg.test.explorer.fixture',reporter~+~=~+~VSCodeReporter)
    ```

    Close the Positron window and confirm runner is gone:
    ```
    .../resources/testing/r.pkg.test.explorer.fixture $ ps -ww -p 97094 -o pid,ppid,command
      PID  PPID COMMAND
    ```


- [x] F Window reload mid-run. Just like scenario E, except for window reload instead of close.
- [x] G1 Immediate cancel. Cancel right after starting, before much happens. Expect: no stuck state; you can start another run.
- [x] G2 -- Re-run after cancel. Cancel a run, then start another. Expect: stale results from the cancelled run are cleared and the new run proceeds normally.

**Review by juliasilge — APPROVED** [review]
This is working well for me! 🙌
